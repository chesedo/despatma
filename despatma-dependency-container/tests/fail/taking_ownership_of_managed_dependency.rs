struct Config {
    port: u32,
}

struct Service;

impl Service {
    fn new(port: u32) -> Self {
        println!("Service started on port {}", port);
        Self
    }
}

#[despatma_dependency_container::dependency_container]
impl DependencyContainer {
    #[Singleton]
    fn config(&self) -> Config {
        Config { port: 8080 }
    }

    // To be able to manage config, it has to be passed around by reference
    fn service(&self, config: Config) -> Service {
        Service::new(config.port)
    }
}

fn main() {
    let container = DependencyContainer::new();
    let _service = container.service();
}
