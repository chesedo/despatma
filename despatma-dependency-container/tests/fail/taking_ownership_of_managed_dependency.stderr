error: This dependency is managed by the container

         = help: Take a reference here instead: &Config

  --> tests/fail/taking_ownership_of_managed_dependency.rs:22:31
   |
22 |     fn service(&self, config: Config) -> Service {
   |                               ^^^^^^

error[E0433]: failed to resolve: use of undeclared type `DependencyContainer`
  --> tests/fail/taking_ownership_of_managed_dependency.rs:28:21
   |
28 |     let container = DependencyContainer::new();
   |                     ^^^^^^^^^^^^^^^^^^^ use of undeclared type `DependencyContainer`
